### PR TITLE
chore(refactor): add global db in memory support - driver builder 2/2 (WPB-8989)

### PR DESCRIPTION
**Pull request by yamilmedina**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

This PR addresses Builder pattern dsl and remove expect/actual for driver creation.

### Solutions

Apply `dsl` builder pattern, suggested in prev. PR 
- #2747

### CODE REVIEW

File: persistence/src/androidMain/kotlin/com/wire/kalium/persistence/db/PlatformDatabaseData.kt [hunk 1]
```diff
@@ -27,3 +33,32 @@ import android.content.Context
...
+    context: Context,
+    dbName: String,
+    isEncryptionEnabled: Boolean,
+    passphrase: ByteArray? = null,
+    schema: SqlSchema<QueryResult.Value<Unit>>,
+    config: DriverConfigurationBuilder.() -> Unit = {}
+): SqlDriver {
+    val driverConfiguration = DriverConfigurationBuilder().apply(config)
+    val enableWAL = driverConfiguration.isWALEnabled
+    return if (isEncryptionEnabled) {
```
Comment (vitorhugods) [inline persistence/src/androidMain/kotlin/com/wire/kalium/persistence/db/PlatformDatabaseData.kt:48]: `isEncryptionEnabled` can be removed 😄 

```suggestion
    passphrase: ByteArray? = null,
    schema: SqlSchema<QueryResult.Value<Unit>>,
    config: DriverConfigurationBuilder.() -> Unit = {}
): SqlDriver {
    val driverConfiguration = DriverConfigurationBuilder().apply(config)
    val enableWAL = driverConfiguration.isWALEnabled
    return if (passphrase != null) {
```

**Review by vitorhugods — APPROVED** [review]
Noice 🧑‍🍳 

Just one tiny suggestion

File: persistence/src/jvmMain/kotlin/com/wire/kalium/persistence/db/UserDatabase.kt [hunk 2]
```diff
@@ -67,9 +69,10 @@ actual fun userDatabaseDriverByPath(
     path: String,
     passphrase: UserDBSecret?,
     enableWAL: Boolean
-): SqlDriver = DriverBuilder()
-    .withWALEnabled(enableWAL)
-    .build(path)
+): SqlDriver = databaseDriver(JdbcSqliteDriver.IN_MEMORY) {
+    isWALEnabled = enableWAL
+    areForeignKeyConstraintsEnforced = true
+}
```
Comment (alexandreferris) [inline persistence/src/jvmMain/kotlin/com/wire/kalium/persistence/db/UserDatabase.kt:75]: Does it make sense changing the name of the method? as its state we are getting the DB driver by path and this is not true anymore? 🤔

**Review by alexandreferris — COMMENTED** [review]
Great work!

I just have one comment on a method return change 🤔

File: persistence/src/jvmMain/kotlin/com/wire/kalium/persistence/db/UserDatabase.kt (hunk 2 above)
Comment (yamilmedina) [reply persistence/src/jvmMain/kotlin/com/wire/kalium/persistence/db/UserDatabase.kt:75]: good catch, actually we should use it :O

File: persistence/src/androidMain/kotlin/com/wire/kalium/persistence/db/PlatformDatabaseData.kt (hunk 1 above)
Comment (yamilmedina) [reply persistence/src/androidMain/kotlin/com/wire/kalium/persistence/db/PlatformDatabaseData.kt]: Good idea, just needed to lift the check for android, depending on kalium configs.

**Review by alexandreferris — APPROVED** [review]

**Pull request merged**